text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+Another way to state the problem is that apps and addons all _push_ whatever code they want into the final built app. Whereas ES modules can _pull_ each other into the build as needed.
+
+### Play nice with NPM Conventions
+
+The ECMA module spec by itself doesn’t try to define a module resolution algorithm. But the overwhelmingly most popular convention is the [node_modules resolution algorithm](https://nodejs.org/api/all.html#modules_all_together).
+
+Ember addons do respect node_module resolution for build-time code, but they do not respect it for runtime code. This is an unhelpful distinction.
+
+### Verbose, Static Javascript as a Compiler Target
+
+Ember’s strong conventions mean that many kinds of dependencies can be inferred (including _statically_ inferred) without requiring the developer to laboriously manage them. This is a good thing and I believe the current fad in the wider Javascript ecosystem for making developers hand-write verbose static imports for everything confuses the benefits of having static analysis (which is good) with the benefits of hand-managing those static imports (which is unnecessary cognitive load when you have clear conventions and a compiler).
+
+This design is about compiling today’s idiomatic Ember code into more “vanilla” patterns that leverage ES modules, node_modules resolution, and spec-compliant static and dynamic `import` to express the structure of an Ember application in a much more “vanilla Javascript” way.
+
+This compile step lets us separate the authoring format (which isn’t changing in any significant way in this RFC) from the packaging format (which can be more verbose and static than we would want in an authoring format).
+
+# Detailed design
+
+## Definitions
+
+**package**: every addon and app is a package. Usually synonymous with “NPM package”, but we also include in-repo packages. The most important fact about a package is that it’s often the boundary around code that comes from a particular author, team, or organization, so coordination across packages is a more sensitive design problem than coordination within apps.
+
+**app**: a package used at the root of a project.
+
+**addon**: a package not used at the root of a project. Will be an **allowed dependency** of either an **app** or an **addon**.
+
+**allowed dependency**: For **addons**, the **allowed dependencies** are the `dependencies` and `peerDependencies` in `package.json` plus any in-repo addons. For **apps**, the **allowed dependencies** are the `dependencies`, `peerDependencies`, and `devDependencies` in `package.json` plus any in-repo addons.
+
+**Ember package metadata**: the `ember-addon` section inside `package.json`. This already exists in v1, we’re going to extend it.
+
+**v2 package**: a package with `package.json` like:
+
+    "keywords": [ "ember-addon" ],
+    "ember-addon": {
```

Comment:
I agree we can adopt `ember` instead of `ember-addon`. 

Using a real market package would have distinct disadvantages. NPM would really download  it, and it would get traversed over and over in the graph because it would be a leaf beneath every addon. And we really *don't* want people to import any code out of it, because that's the kind of implementation-defined behavior this RFC is trying to move away from.

As for making the version more like a package string, I don't think that's a good fit. If an integer turns out to be insufficient, the next step up would be a capability set, not a semver string. This is much more like the capabilities on a component manager than the version of a package.